File: snapshot-lib/src/jvmTest/kotlin/com/diffplug/selfie/SnapshotValueReaderTest.kt
/*
 * Copyright (C) 2023 DiffPlug
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.diffplug.selfie

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.matchers.shouldBe
import kotlin.test.Test

class SnapshotValueReaderTest {
  @Test
  fun noEscapingNeeded() {
    val reader =
        SnapshotValueReader.of(
            """
            ╔═ 00_empty ═╗
            ╔═ 01_singleLineString ═╗
            this is one line
            ╔═ 01a_singleLineLeadingSpace ═╗
             the leading space is significant
            ╔═ 01b_singleLineTrailingSpace ═╗
            the trailing space is significant 
            ╔═ 02_multiLineStringTrimmed ═╗
            Line 1
            Line 2
            ╔═ 03_multiLineStringTrailingNewline ═╗
            Line 1
            Line 2

            ╔═ 04_multiLineStringLeadingNewline ═╗

            Line 1
            Line 2
            ╔═ 05_notSureHowKotlinMultilineWorks ═╗
            """
                .trimIndent())
    reader.peekKey() shouldBe "00_empty"
    reader.peekKey() shouldBe "00_empty"
    reader.nextValue().valueString() shouldBe ""
    reader.peekKey() shouldBe "01_singleLineString"
    reader.peekKey() shouldBe "01_singleLineString"
    reader.nextValue().valueString() shouldBe "this is one line"
    reader.peekKey() shouldBe "01a_singleLineLeadingSpace"
    reader.nextValue().valueString() shouldBe " the leading space is significant"
    reader.peekKey() shouldBe "01b_singleLineTrailingSpace"
    reader.nextValue().valueString() shouldBe "the trailing space is significant "
    reader.peekKey() shouldBe "02_multiLineStringTrimmed"
    reader.nextValue().valueString() shouldBe "Line 1\nLine 2"
    // note that leading and trailing newlines in the snapshots are significant
    // this is critical so that snapshots can accurately capture the exact number of newlines
    reader.peekKey() shouldBe "03_multiLineStringTrailingNewline"
    reader.nextValue().valueString() shouldBe "Line 1\nLine 2\n"
    reader.peekKey() shouldBe "04_multiLineStringLeadingNewline"
    reader.nextValue().valueString() shouldBe "\nLine 1\nLine 2"
    reader.peekKey() shouldBe "05_notSureHowKotlinMultilineWorks"
    reader.nextValue().valueString() shouldBe ""
  }

  @Test
  fun invalidNames() {
    shouldThrow<ParseException> { SnapshotValueReader.of("╔═name ═╗").peekKey() }
        .let { it.message shouldBe "L1:Expected to start with '╔═ '" }
    shouldThrow<ParseException> { SnapshotValueReader.of("╔═ name═╗").peekKey() }
        .let { it.message shouldBe "L1:Expected to contain ' ═╗'" }
    shouldThrow<ParseException> { SnapshotValueReader.of("╔═  name ═╗").peekKey() }
        .let { it.message shouldBe "L1:Leading spaces are disallowed: ' name'" }
    shouldThrow<ParseException> { SnapshotValueReader.of("╔═ name  ═╗").peekKey() }
        .let { it.message shouldBe "L1:Trailing spaces are disallowed: 'name '" }
    SnapshotValueReader.of("╔═ name ═╗ comment okay").peekKey() shouldBe "name"
    SnapshotValueReader.of("╔═ name ═╗okay here too").peekKey() shouldBe "name"
    SnapshotValueReader.of("╔═ name ═╗ okay  ╔═ ═╗ (it's the first ' ═╗' that counts)")
        .peekKey() shouldBe "name"
  }

  @Test
  fun escapeCharactersInName() {
    val reader =
        SnapshotValueReader.of(
            """
            ╔═ test with \∕slash\∕ in name ═╗
            ╔═ test with \(square brackets\) in name ═╗
            ╔═ test with \\backslash\\ in name ═╗
            ╔═ test with\nnewline\nin name ═╗
            ╔═ test with \ttab\t in name ═╗
            ╔═ test with \┌\─ ascii art \─\┐ in name ═╗
            """
                .trimIndent())
    reader.peekKey() shouldBe "test with /slash/ in name"
    reader.nextValue().valueString() shouldBe ""
    reader.peekKey() shouldBe "test with [square brackets] in name"
    reader.nextValue().valueString() shouldBe ""
    reader.peekKey() shouldBe """test with \backslash\ in name"""
    reader.nextValue().valueString() shouldBe ""
    reader.peekKey() shouldBe
        """
        test with
        newline
        in name
         """
            .trimIndent()
    reader.nextValue().valueString() shouldBe ""
    reader.peekKey() shouldBe "test with \ttab\t in name"
    reader.nextValue().valueString() shouldBe ""
    reader.peekKey() shouldBe "test with ╔═ ascii art ═╗ in name"
    reader.nextValue().valueString() shouldBe ""
  }

  @Test
  fun escapeCharactersInBody() {
    val reader =
        SnapshotValueReader.of(
            """
          ╔═ ascii art okay ═╗
           ╔══╗
          ╔═ escaped iff on first line ═╗
          𐝁══╗
          ╔═ body escape characters ═╗
          𐝃𐝁𐝃𐝃 linear a is dead
        """
                .trimIndent())
    reader.peekKey() shouldBe "ascii art okay"
    reader.nextValue().valueString() shouldBe """ ╔══╗"""
    reader.peekKey() shouldBe "escaped iff on first line"
    reader.nextValue().valueString() shouldBe """╔══╗"""
    reader.peekKey() shouldBe "body escape characters"
    reader.nextValue().valueString() shouldBe """𐝁𐝃 linear a is dead"""
  }

  @Test
  fun skipValues() {
    val testContent =
        """
            ╔═ 00_empty ═╗
            ╔═ 01_singleLineString ═╗
            this is one line
            ╔═ 02_multiLineStringTrimmed ═╗
            Line 1
            Line 2
            ╔═ 05_notSureHowKotlinMultilineWorks ═╗
            """
            .trimIndent()
    assertKeyValueWithSkip(testContent, "00_empty", "")
    assertKeyValueWithSkip(testContent, "01_singleLineString", "this is one line")
    assertKeyValueWithSkip(testContent, "02_multiLineStringTrimmed", "Line 1\nLine 2")
  }
  private fun assertKeyValueWithSkip(input: String, key: String, value: String) {
    val reader = SnapshotValueReader.of(input)
    while (reader.peekKey() != key) {
      reader.skipValue()
    }
    reader.peekKey() shouldBe key
    reader.nextValue().valueString() shouldBe value
    while (reader.peekKey() != null) {
      reader.skipValue()
    }
  }
}

File: python/selfie-lib/tests/SnapshotValueReader_test.py
import pytest
from selfie_lib import SnapshotValueReader, ParseException


class TestSnapshotValueReader:
    def test_no_escaping_needed(self):
        reader = SnapshotValueReader.of(
            """
            ╔═ 00_empty ═╗
            ╔═ 01_singleLineString ═╗
            this is one line
            ╔═ 01a_singleLineLeadingSpace ═╗
             the leading space is significant
            ╔═ 01b_singleLineTrailingSpace ═╗
            the trailing space is significant
            ╔═ 02_multiLineStringTrimmed ═╗
            Line 1
            Line 2
            ╔═ 03_multiLineStringTrailingNewline ═╗
            Line 1
            Line 2

            ╔═ 04_multiLineStringLeadingNewline ═╗

            Line 1
            Line 2
            ╔═ 05_notSureHowKotlinMultilineWorks ═╗
        """.strip()
        )
        assert reader.peek_key() == "00_empty"
        assert reader.next_value().value_string() == ""
        assert reader.peek_key() == "01_singleLineString"
        assert reader.next_value().value_string() == "this is one line"
        assert reader.peek_key() == "01a_singleLineLeadingSpace"
        assert reader.next_value().value_string() == " the leading space is significant"
        assert reader.peek_key() == "01b_singleLineTrailingSpace"
        assert (
            reader.next_value().value_string() == "the trailing space is significant "
        )
        assert reader.peek_key() == "02_multiLineStringTrimmed"
        assert reader.next_value().value_string() == "Line 1\nLine 2"
        assert reader.peek_key() == "03_multiLineStringTrailingNewline"
        assert reader.next_value().value_string() == "Line 1\nLine 2\n"
        assert reader.peek_key() == "04_multiLineStringLeadingNewline"
        assert reader.next_value().value_string() == "\nLine 1\nLine 2"
        assert reader.peek_key() == "05_notSureHowKotlinMultilineWorks"
        assert reader.next_value().value_string() == ""

    def test_invalid_names(self):
        with pytest.raises(ParseException) as exc_info:
            SnapshotValueReader.of("╔═name ═╗").peek_key()
        assert "Expected to start with '╔═ '" in str(exc_info.value)

        with pytest.raises(ParseException) as exc_info:
            SnapshotValueReader.of("╔═ name═╗").peek_key()
        assert "Expected to contain ' ═╗'" in str(exc_info.value)

        with pytest.raises(ParseException) as exc_info:
            SnapshotValueReader.of("╔═  name ═╗").peek_key()
        assert "Leading spaces are disallowed: ' name'" in str(exc_info.value)

        with pytest.raises(ParseException) as exc_info:
            SnapshotValueReader.of("╔═ name  ═╗").peek_key()
        assert "Trailing spaces are disallowed: 'name '" in str(exc_info.value)

        assert SnapshotValueReader.of("╔═ name ═╗ comment okay").peek_key() == "name"
        assert SnapshotValueReader.of("╔═ name ═╗okay here too").peek_key() == "name"
        assert (
            SnapshotValueReader.of(
                "╔═ name ═╗ okay  ╔═ ═╗ (it's the first ' ═╗' that counts)"
            ).peek_key()
            == "name"
        )

    def test_escape_characters_in_name(self):
        reader = SnapshotValueReader.of(
            """
            ╔═ test with \\(square brackets\\) in name ═╗
            ╔═ test with \\\\backslash\\\\ in name ═╗
            ╔═ test with\\nnewline\\nin name ═╗
            ╔═ test with \\ttab\\t in name ═╗
            ╔═ test with \\┌\\─ ascii art \\─\\┐ in name ═╗
        """.strip()
        )
        assert reader.peek_key() == "test with (square brackets) in name"
        assert reader.next_value().value_string() == ""
        assert reader.peek_key() == "test with \\backslash\\ in name"
        assert reader.next_value().value_string() == ""
        assert reader.peek_key().strip() == "test with\nnewline\nin name"
        assert reader.next_value().value_string() == ""
        assert reader.peek_key() == "test with \ttab\t in name"
        assert reader.next_value().value_string() == ""
        assert reader.peek_key() == "test with ╔═ ascii art ═╗ in name"
        assert reader.next_value().value_string() == ""

    def assert_key_value_with_skip(self, key, expected_value):
        reader = SnapshotValueReader.of(self)
        while reader.peek_key() != key:
            reader.skip_value()
        assert reader.peek_key() == key
        assert reader.next_value().value_string() == expected_value
        while reader.peek_key() is not None:
            reader.skip_value()

    def test_skip_values(self):
        test_content = """
            ╔═ 00_empty ═╗
            ╔═ 01_singleLineString ═╗
            this is one line
            ╔═ 02_multiLineStringTrimmed ═╗
            Line 1
            Line 2
            ╔═ 05_notSureHowKotlinMultilineWorks ═╗
        """.strip()
        self.assert_key_value_with_skip(test_content, "00_empty", "")
        self.assert_key_value_with_skip(
            test_content, "01_singleLineString", "this is one line"
        )
        self.assert_key_value_with_skip(
            test_content, "02_multiLineStringTrimmed", "Line 1\nLine 2"
        )

    def test_binary():
        reader = SnapshotValueReader(
            """╔═ Apple ═╗ base64 length 3 bytes
    c2Fk
    """
        )
        assert reader.peek_key() == "Apple"
        assert reader.next_value() == b"sad"
